src/ml_board/backend/restful_api/data_access.py

```diff
@@ -236,6 +240,70 @@ def get_experiment_config(self, grid_search_id: str, experiment_id: str, config_
         else:
             raise InvalidPathError(f"File path {requested_full_path} is not safe.")
 
+    def get_checkpoint_dict_epoch(self, grid_search_id: str, experiment_id: str, epoch: str):
+        """
+        Fetch all checkpoint resource pickle files from event storage
+        given the epoch, experiment ID & grid search ID.
+
+        :params:
+             grid_search_id (str): Grid Search ID
+             experiment_id (str): Experiment ID
+             epoch (str): Epoch number
+
+        :returns: List of Checkpoint files
+        """
+        requested_full_path = os.path.realpath(
+            os.path.join(self.top_level_logging_path, str(grid_search_id), str(experiment_id), str(epoch))
+        )
+
+        if FileDataAccess.is_safe_path(base_dir=self.top_level_logging_path, requested_path=requested_full_path):
+            response = []
+            checkpoints = []
+            files = FileDataAccess.get_checkpoint_files(requested_full_path, base_path=self.top_level_logging_path)
+            for file_num in range(len(files)):
+                split = os.path.normpath(files[file_num]).split(os.sep)
+                checkpoints.append(os.path.basename(split[3]).split(".")[0])
+                if file_num == len(files) - 1:
+                    response.append({"experiment_id": experiment_id, "epoch": epoch, "checkpoints": checkpoints})
+            return response
+
+        else:
+            raise InvalidPathError(f"File path {requested_full_path} is not safe.")
+
+    def get_checkpoint_list(self, grid_search_id: str, experiment_id):
+        """
+        `Fetch checkpoint resource pickle file given the experiment ID & grid search ID from event storage.
+
+        :params:
+             grid_search_id (str): Grid Search ID
+             experiment_id (str): Experiment ID
+
+        :returns: list of checkpoints
+        """
+        requested_full_path = os.path.realpath(os.path.join(self.top_level_logging_path, str(grid_search_id), str(experiment_id)))
+
+        if FileDataAccess.is_safe_path(base_dir=self.top_level_logging_path, requested_path=requested_full_path):
+            response = []
+            checkpoints = []
+            files = FileDataAccess.get_checkpoint_files(requested_full_path, base_path=self.top_level_logging_path)
+            last_epoch = os.path.normpath(files[0]).split(os.sep)[2]
```

Comment:
I don't understand this line. 
files[0] should give a single file path. you split then on / and take the second element in the list. 
Since the file path can be /foo/bar/<grid_search_id>/<experiment_id><epoch>checkpoint_file.pickle, -2 as the index would make more sense.

Reply:
In an experiment, we can have multiple epochs. I needed a reference point to understand when I move to another epoch so I can create different checkpoint lists for different epochs in the same experiment. 

the files[0] is the first checkpoint file pat. I want to save the epoch in the checkpoint path for later use hence os.path.normpath(files[0]).split(os.sep)[2] (where index 2 is the epoch name FYI: can be changed to -2 same thing)

The file path is rigid and will be grid-search-id/experiment-id/epoch/checkpoint.pickle